File: src/main/java/com/heybys/oddments/base/domain/DomainEntity.java
package com.heybys.oddments.base.domain;

public abstract class DomainEntity<T extends DomainEntity<T, I>, I> {

    public abstract I getId();

    @Override
    public boolean equals(Object other) {
        if (other == null) {
            return false;
        }

        if (!other.getClass().equals(getClass())) {
            return false;
        }

        return equals((T) other);
    }

    public boolean equals(T other) {
        if (other == null) {
            return false;
        }

        if (getId() == null) {
            return false;
        }

        if (other.getClass().equals(getClass())) {
            return getId().equals(other.getId());
        }

        return super.equals(other);
    }

    @Override
    public int hashCode() {
        return getId() == null ? 0 : getId().hashCode();
    }
}

File: src/main/kotlin/com/heybys/oddments/base/domain/DomainEntity.kt
package com.heybys.oddments.base.domain

abstract class DomainEntity<T : DomainEntity<T, I>, I> : BaseEntity() {

    abstract fun getId(): I?

    override fun equals(other: Any?): Boolean {
        if (other == null) return false
        if (other.javaClass != javaClass) return false

        @Suppress("UNCHECKED_CAST")
        return equals(other as T)
    }

    open fun equals(other: T?): Boolean {
        if (other == null) return false
        val id = getId()
        if (id == null) return false

        return if (other.javaClass == javaClass) {
            id == other.getId()
        } else {
            super.equals(other)
        }
    }

    override fun hashCode(): Int {
        return getId()?.hashCode() ?: 0
    }
}
